### PR TITLE
Use a separate node pool to run custom connector jobs

### PR DESCRIPTION
**Pull request by xiaohansong**
## What
https://github.com/airbytehq/airbyte/issues/19684

Custom connectors - user can upload their own docker image as the source or destination of the sync. Such sourceDefinition or destinationDefinition will be classified as custom connectors. Since they are not vetted by airbyte, we want to run it in a separate node pool for safety reasons to avoid failing jobs running with airbyte vetted connectors.

The goal of the PR is to:
1) Identify which activity are involving custom connectors.
2) pass along custom connector bit via temporal and eventually when starting a kube process we know it's a custom connector and thus run it on a different node pool.

## How
Passing the value in launchConfig - launcher will read launchConfig and launch a kube process with correct node selector.

For jobs spawned by Orchestrator, we also need to pass down the env variable to containers.

### CODE REVIEW

File: airbyte-commons-worker/src/main/java/io/airbyte/workers/normalization/DefaultNormalizationRunner.java [hunk 1]
```diff
@@ -138,6 +138,8 @@ private boolean runProcess(final String jobId,
           attempt,
           jobRoot,
           normalizationImageName,
+          // custom connector does not use normalization
```
Comment (xiaohansong) [inline airbyte-commons-worker/src/main/java/io/airbyte/workers/normalization/DefaultNormalizationRunner.java:141]: @davinchia Can you help me confirm if this is true? (seems NormalizationRunner will only be running if it's called out here: https://github.com/airbytehq/airbyte/blob/master/airbyte-commons-worker/src/main/java/io/airbyte/workers/normalization/NormalizationRunnerFactory.java#L18-L38)

File: airbyte-commons-worker/src/main/java/io/airbyte/workers/general/DbtTransformationRunner.java [hunk 2]
```diff
@@ -99,6 +99,7 @@ public boolean transform(final String jobId,
               jobRoot,
               dbtConfig.getDockerImage(),
               false,
```
Comment (xiaohansong) [inline airbyte-commons-worker/src/main/java/io/airbyte/workers/general/DbtTransformationRunner.java:101]: @davinchia or @mfsiega-airbyte I don't have much context on this one - from the comment it seems transformation will only be provided by user, but for custom connector are we allowing them to use Dbt related features?

Comment (pmossman) [reply airbyte-commons-worker/src/main/java/io/airbyte/workers/general/DbtTransformationRunner.java:101]: We don't currently support custom DBT transformations in Cloud (and we deprioritized the work to enable custom transformations on kube this quarter) but I think DBT transformations are only supported in a specific list of destinations, just like with normalization. So for custom destination connectors, I don't think we need to worry about DBT transformations or Normalization (@davinchia should confirm though to be sure)

File: airbyte-commons-worker/src/main/java/io/airbyte/workers/normalization/DefaultNormalizationRunner.java (hunk 1 above)
Comment (pmossman) [reply airbyte-commons-worker/src/main/java/io/airbyte/workers/normalization/DefaultNormalizationRunner.java:141]: I believe this is true, normalization is supported in a specific list of destination connectors, so I don't think we would be able to support normalization for custom destination connectors (at least for V1)

Comment (davinchia) [reply airbyte-commons-worker/src/main/java/io/airbyte/workers/normalization/DefaultNormalizationRunner.java:141]: Parker is correct!

File: airbyte-commons-worker/src/main/java/io/airbyte/workers/general/DbtTransformationRunner.java (hunk 2 above)
Comment (davinchia) [reply airbyte-commons-worker/src/main/java/io/airbyte/workers/general/DbtTransformationRunner.java:101]: yeap, Parker is correct here too!

**Comment by davinchia** [comment]
@xiaohansong apart from the 2 questions, are there specific bits you would like me to review? I think my review is not needed if Parker is going over this - putting it out here if there you want a second eye.

**Review by pmossman — CHANGES_REQUESTED** [review]
@xiaohansong I just finished going through this, great stuff! I made some suggestions and flagged a couple of questions/thoughts but I think this looks pretty close to me. 

@davinchia I think you could just look through my comments and weigh in if you have thoughts, rather than reviewing everything. I think I commented everywhere that is worth taking a look, the rest of the PR is straightforward and/or boilerplate

File: airbyte-commons-worker/src/main/java/io/airbyte/workers/process/DockerProcessFactory.java [hunk 3]
```diff
@@ -83,6 +83,7 @@ public Process create(final String jobType,
                         final int attempt,
                         final Path jobRoot,
                         final String imageName,
+                        final boolean usesIsolatedPool,
```
[the remark this answers is not in the document]
Comment (xiaohansong) [reply airbyte-commons-worker/src/main/java/io/airbyte/workers/process/DockerProcessFactory.java:86]: I didn't know docker has this pool concept as well. Changed to use `usesIsolatedPool`.

File: airbyte-commons-worker/src/main/java/io/airbyte/workers/sync/ReplicationLauncherWorker.java [hunk 4]
```diff
@@ -54,7 +54,8 @@ public ReplicationLauncherWorker(final UUID connectionId,
         activityContext,
         serverPort,
         temporalUtils,
-        workerConfigs);
+        workerConfigs,
+        sourceLauncherConfig.getIsCustomConnector() || destinationLauncherConfig.getIsCustomConnector());
```
[the remark this answers is not in the document]
Comment (xiaohansong) [reply airbyte-commons-worker/src/main/java/io/airbyte/workers/sync/ReplicationLauncherWorker.java:58]: Fixed!

File: airbyte-container-orchestrator/src/main/java/io/airbyte/container_orchestrator/ReplicationJobOrchestrator.java [hunk 5]
```diff
@@ -106,15 +107,17 @@ public Optional<String> runJob() throws Exception {
...
 
     log.info("Setting up destination launcher...");
     final IntegrationLauncher destinationLauncher = new AirbyteIntegrationLauncher(
         destinationLauncherConfig.getJobId(),
         Math.toIntExact(destinationLauncherConfig.getAttemptId()),
         destinationLauncherConfig.getDockerImage(),
         processFactory,
-        syncInput.getDestinationResourceRequirements());
+        syncInput.getDestinationResourceRequirements(),
+        destinationLauncherConfig.getIsCustomConnector());
```
[the remark this answers is not in the document]
Comment (xiaohansong) [reply airbyte-container-orchestrator/src/main/java/io/airbyte/container_orchestrator/ReplicationJobOrchestrator.java:120]: yup done.

File: airbyte-workers/src/main/java/io/airbyte/workers/config/WorkerConfigurationBeanFactory.java [hunk 6]
```diff
@@ -85,6 +85,18 @@ public Map<String, String> specJobKubeAnnotations(@Value("${airbyte.worker.spec.
     return splitKVPairsFromEnvString(kubeAnnotations);
   }
 
+  @Singleton
+  @Named("customNodeSelectors")
```
[the remark this answers is not in the document]
Comment (xiaohansong) [reply airbyte-workers/src/main/java/io/airbyte/workers/config/WorkerConfigurationBeanFactory.java:89]: decide to use `useIsolatedNodeSelector` to keep align with other names (such as specJobKubeNodeSelectors). Since this is to define the actual kube selector I think we should just include `node` in the name

File: airbyte-commons-worker/src/main/java/io/airbyte/workers/normalization/DefaultNormalizationRunner.java (hunk 1 above)
Comment (xiaohansong) [reply airbyte-commons-worker/src/main/java/io/airbyte/workers/normalization/DefaultNormalizationRunner.java:141]: Perfect, thanks!

File: airbyte-commons-worker/src/main/java/io/airbyte/workers/WorkerConfigs.java [hunk 7]
```diff
@@ -41,6 +42,7 @@ public WorkerConfigs(final Configs configs) {
             .withMemoryLimit(configs.getJobMainContainerMemoryLimit()),
         configs.getJobKubeTolerations(),
         configs.getJobKubeNodeSelectors(),
+        configs.getUseCustomKubeNodeSelector() ? configs.getIsolatedJobKubeNodeSelectors() : configs.getJobKubeNodeSelectors(),
```
[the remark this answers is not in the document]
Comment (xiaohansong) [reply airbyte-commons-worker/src/main/java/io/airbyte/workers/WorkerConfigs.java:45]: I agree. Changed!

File: airbyte-commons-worker/src/main/java/io/airbyte/workers/process/KubeProcessFactory.java [hunk 8]
```diff
@@ -107,6 +108,8 @@ public Process create(
 
       final var allLabels = getLabels(jobId, attempt, customLabels);
 
+      final var nodeSelectors = usesIsolatedPool ? workerConfigs.getWorkerIsolatedKubeNodeSelectors() : workerConfigs.getworkerKubeNodeSelectors();
```
[the remark this answers is not in the document]
Comment (xiaohansong) [reply airbyte-commons-worker/src/main/java/io/airbyte/workers/process/KubeProcessFactory.java:111]: it can be empty if the flag was set to false - if they are using custom connectors but didn't turn on the useIsolatedNodePool flag, they might hit this

**Comment by xiaohansong** [comment]
@pmossman when you get a chance can you take another look at this PR? thanks!

File: airbyte-commons-worker/src/main/java/io/airbyte/workers/sync/LauncherWorker.java [hunk 9]
```diff
@@ -176,13 +176,19 @@ public OUTPUT run(final INPUT input, final Path jobRoot) throws WorkerException
           log.info("Creating " + podName + " for attempt number: " + jobRunConfig.getAttemptId());
           killRunningPodsForConnection();
 
+          // custom connectors run in an isolated node pool from airbyte-supported connectors
+          // to reduce the blast radius of any problems with custom connector code.
+          final var nodeSelectors =
+              isCustomConnector ? workerConfigs.getWorkerIsolatedKubeNodeSelectors().orElse(workerConfigs.getworkerKubeNodeSelectors())
```
Comment (pmossman) [inline airbyte-commons-worker/src/main/java/io/airbyte/workers/sync/LauncherWorker.java:182]: Similar comment here, I think it's more dangerous to fall back on our primary node selectors, and perhaps we should throw an exception instead.

File: airbyte-server/src/main/java/io/airbyte/server/handlers/SourceDefinitionsHandler.java [hunk 10]
```diff
@@ -285,16 +289,6 @@ public void deleteSourceDefinition(final SourceDefinitionIdRequestBody sourceDef
     configRepository.writeStandardSourceDefinition(persistedSourceDefinition);
   }
 
-  public void deleteCustomSourceDefinition(final SourceDefinitionIdWithWorkspaceId sourceDefinitionIdWithWorkspaceId)
```
Comment (pmossman) [inline airbyte-server/src/main/java/io/airbyte/server/handlers/SourceDefinitionsHandler.java]: is this change supposed to be part of this PR? Or did it sneak in from your other branch for consolidating API endpoints?

File: airbyte-server/src/test/java/io/airbyte/server/handlers/DestinationDefinitionsHandlerTest.java [hunk 11]
```diff
@@ -336,52 +336,6 @@ void testGetDefinitionWithGrantForWorkspace() throws JsonValidationException, Co
     assertEquals(expectedDestinationDefinitionRead, actualDestinationDefinitionRead);
   }
 
-  @Test
-  @DisplayName("createDestinationDefinition should correctly create a destinationDefinition")
```
Comment (pmossman) [inline airbyte-server/src/test/java/io/airbyte/server/handlers/DestinationDefinitionsHandlerTest.java]: similar question here, is this change sneaking in from your other PR?

File: airbyte-server/src/test/java/io/airbyte/server/handlers/SourceDefinitionsHandlerTest.java [hunk 12]
```diff
@@ -324,54 +324,6 @@ void testGetDefinitionWithGrantForWorkspace() throws JsonValidationException, Co
     assertEquals(expectedSourceDefinitionRead, actualSourceDefinitionRead);
   }
 
-  @Test
```
Comment (pmossman) [inline airbyte-server/src/test/java/io/airbyte/server/handlers/SourceDefinitionsHandlerTest.java:327]: similar question here, is this change sneaking in from your other PR?

File: airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/activities/GenerateInputActivityImpl.java [hunk 13]
```diff
@@ -111,7 +111,9 @@ public GeneratedJobInput getSyncWorkflowInput(final SyncInput input) {
           .withAttemptId((long) attempt)
           .withDockerImage(config.getDestinationDockerImage())
           .withProtocolVersion(config.getDestinationProtocolVersion())
-          .withIsCustomConnector(config.getIsDestinationCustomConnector());
+          .withIsCustomConnector(config.getIsDestinationCustomConnector())
```
Comment (pmossman) [inline airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/activities/GenerateInputActivityImpl.java:114]: These changes look unrelated, maybe due to a merge/rebase with master?

**Review by pmossman — APPROVED** [review]
@xiaohansong I left a couple more comments but overall this looks good to me! I'll approve the PR assuming we don't disagree about any of my comments, let me know if you want to sync about any of them!

File: airbyte-commons-worker/src/main/java/io/airbyte/workers/process/KubeProcessFactory.java [hunk 14]
```diff
@@ -108,7 +108,11 @@ public Process create(
 
       final var allLabels = getLabels(jobId, attempt, customLabels);
 
-      final var nodeSelectors = usesIsolatedPool ? workerConfigs.getWorkerIsolatedKubeNodeSelectors() : workerConfigs.getworkerKubeNodeSelectors();
+      // If using isolated pool, check workerConfigs has isolated pool set. If not set, fall back to use
+      // regular node pool.
+      final var nodeSelectors =
+          usesIsolatedPool ? workerConfigs.getWorkerIsolatedKubeNodeSelectors().orElse(workerConfigs.getworkerKubeNodeSelectors())
```
[the remark this answers is not in the document]
Comment (xiaohansong) [reply airbyte-commons-worker/src/main/java/io/airbyte/workers/process/KubeProcessFactory.java:114]: good call! fixed

File: airbyte-commons-worker/src/main/java/io/airbyte/workers/sync/LauncherWorker.java (hunk 9 above)
Comment (xiaohansong) [reply airbyte-commons-worker/src/main/java/io/airbyte/workers/sync/LauncherWorker.java:182]: I added this check in `WorkerConfigurationBeanFactory.java` instead - because 1) workerConfigs does not carry `useIsolatedPool` flag, 2) should throw this error as early as possible

File: airbyte-server/src/main/java/io/airbyte/server/handlers/SourceDefinitionsHandler.java (hunk 10 above)
Comment (xiaohansong) [reply airbyte-server/src/main/java/io/airbyte/server/handlers/SourceDefinitionsHandler.java]: ah true

File: airbyte-server/src/test/java/io/airbyte/server/handlers/SourceDefinitionsHandlerTest.java (hunk 12 above)
Comment (xiaohansong) [reply airbyte-server/src/test/java/io/airbyte/server/handlers/SourceDefinitionsHandlerTest.java:327]: ditto

File: airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/activities/GenerateInputActivityImpl.java (hunk 13 above)
Comment (xiaohansong) [reply airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/activities/GenerateInputActivityImpl.java:114]: yep!

File: airbyte-commons-worker/src/main/java/io/airbyte/workers/sync/LauncherWorker.java (hunk 9 above)
Comment (pmossman) [reply airbyte-commons-worker/src/main/java/io/airbyte/workers/sync/LauncherWorker.java:182]: Nice!

File: airbyte-workers/src/main/java/io/airbyte/workers/config/WorkerConfigurationBeanFactory.java [hunk 15]
```diff
@@ -228,6 +228,9 @@ public WorkerConfigs checkWorkerConfigs(
                                           @Value("${airbyte.worker.job.kube.images.busybox}") final String busyboxImage,
                                           @Value("${airbyte.worker.job.kube.images.curl}") final String curlImage,
                                           @Named("jobDefaultEnvMap") final Map<String, String> jobDefaultEnvMap) {
+    if (useIsolatedNodeSelector && isolatedNodeSelectors.isEmpty()) {
```
Comment (pmossman) [inline airbyte-workers/src/main/java/io/airbyte/workers/config/WorkerConfigurationBeanFactory.java:231]: since this gets repeated so many times, I'd probably move it into a private method that takes in `useIsolatedNodeSelector` and `isolatedNodeSelectors` as arguments

File: airbyte-commons-worker/src/main/java/io/airbyte/workers/process/KubeProcessFactory.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (xiaohansong) [reply airbyte-commons-worker/src/main/java/io/airbyte/workers/process/KubeProcessFactory.java:114]: Realized this is not true - usesIsolatedPool here is actually passed by callers (isCustomConnector flag). Thus for OSS users who do not run them in a separate pool, usesIsolatedPool might still be true because they are custom connectors. I should rename this to isCustomConnectors instead.

[the remark this answers is not in the document]
Comment (pmossman) [reply airbyte-commons-worker/src/main/java/io/airbyte/workers/process/KubeProcessFactory.java]: Oh good call, OSS users should be able to use custom connectors without an isolated pool indeed

**Comment by xiaohansong** [comment]
Connector base build has been failing for unrelated reasons. Submitting this.
